### PR TITLE
Stricter repo parsing

### LINKED ISSUE #10: [Feature Request] Support parsing `blob` and `tree` urls
Can you please add support for "blob" urls like
"https://github.com/jessfraz/dockerfiles/blob/master/buttslock/buttslock.py"

and "tree" urls like
"https://github.com/jessfraz/dockerfiles/tree/master/buttslock"

### LINKED ISSUE #18: URL and repo regex should be non-greedy (maybe)
Hello, 
I think there is a problem in this `regex` or maybe all of them:
```python
import re

pattern = r'https://(?P<domain>.+)/(?P<owner>.+)/(?P<repo>.+?)(?:\.git)?$'
url = 'https://github.com/nephila/giturlparse/blob/master/giturlparse/platforms/github.py'

match = re.match(pattern, url).groupdict()
```
I suppose, this is an invalid URL, for simple reason `git clone` would not accept it; however, when we check the match object, things are a bit strange,:
```
{'domain': 'github.com/nephila/giturlparse/blob/master/giturlparse',
 'owner': 'platforms',
 'repo': 'github.py'}
```
If we make them non-greedy, it would a little better:
```python
pattern = r'https://(?P<domain>.+?)/(?P<owner>.+?)/(?P<repo>.+?)(?:\.git)?$'
match = re.match(pattern, url).groupdict()
```
```
{'domain': 'github.com',
 'owner': 'nephila',
 'repo': 'giturlparse/blob/master/giturlparse/platforms/github.py'}
```
* I don't know whether I'm right or not, but I think it's not normal to see a `/` in repo field, is it?
If it is not, so you guys should return `False` for `valid` attribute, am I right?
* Is there any harm to make all the patterns non-greedy? Is there any circumstances that there would be '/' in the repo field?

@yakky

### PR DESCRIPTION
**Pull request by yakky**
* Domain must not include a slash
* Org must not include a slash on github

Fix #18 
Fix #10

### CODE REVIEW

File: giturlparse/tests/parse.py [hunk 1]
```diff
@@ -112,6 +112,15 @@
...
         'repo': 'Repo',
 
+        'protocol': 'https',
+        'platform': 'gitlab'
+    })),
+    ('HTTPS', ('https://gitlab.com/nephila/giturlparse/blob/master/giturlparse/github.py', {
+        'host': 'gitlab.com',
+        'user': 'git',
+        'owner': 'nephila/giturlparse/blob/master/giturlparse',
+        'repo': 'github.py',
```
Comment (jayvdb) [inline giturlparse/tests/parse.py]: If I understand correctly, previously github.py would end up in the `host` or something like that.

So while this is a bad result for `repo`, it is less bad than before this PR, and at least it is working much better for github.

IMO, "owner" shouldnt contain a `/` on GitLab either.  Subgroups should be accessed using a new member named something like `path` / `pathname`  c.f. https://github.com/retr0h/git-url-parse/pull/37   Or departing from the `urlparse` tuple, a member called `groups` which is a list of strings.

Comment (yakky) [reply giturlparse/tests/parse.py]: @jayvdb this is as far as I am comfortable going with supporting the different system details

what would you think?

Comment (jayvdb) [reply giturlparse/tests/parse.py]: Up to you.  fwiw, https://github.com/coala/git-url-parse has **just** been transferred to the @coala group, which means we can put our orgs needs into that codebase.   Which then leads to how do we cooperate/compete for the name? c.f. https://github.com/coala/git-url-parse/issues/43

Comment (yakky) [reply giturlparse/tests/parse.py]: Sorry if the wording sounded a bit rude, it was not meant. It's just that regexps are a PITA to maintain beyond a certain complexity :)
I also forgot to mention that I updated the PR based on your suggestions, and this PR should be able to support more gitlab features in a more natural way.

Regarding git-url-parse, it certainly isn't super smart to have two competing and very similar implementation that covers this use case, thus definitely let's talk about this. I will answer in the linked issue to stay on topic

File: giturlparse/result.py [hunk 2]
```diff
@@ -51,11 +53,28 @@ def user(self):
...
+    def groups(self):
+        if self.groups_path:
+            return self.groups_path.split('/')
+        else:
+            return []
+
+    @property
+    def path(self):
+        if self.raw_path:
+            return self.raw_path.replace('/blob/', '').strip('/')
```
Comment (jayvdb) [inline giturlparse/result.py]: I dont see the benefit of removing only `blob` (and not other REST 'commands' like 'tree'), but not also remove the branch name.  Keeping the branch name in the `path` makes it like SVN, but for most VCS the branch and the rest of the path are separate and are used in different commands/parts of the workflow.  If they are combined here, people will usually need to split it before they can use it.

If the branch name is also removed, `path` becomes synonymous with the file path that is found in a local clone.

Exposing the `branch` as a separate attribute would be useful, but can be done later.

fwiw, git-url-parse uses `pathname` for what you have exposed as `raw_path`, to correspond with urlparse()'s `pathname`.  I dont mind about this, as a compatibility shim can do simple mapping like this when the two libraries are close enough that a shim is sufficient.

Comment (yakky) [reply giturlparse/result.py]: Regarding the capturing group name, I don't see any reason not to match a more sensible name, I will change that

Regarding the other rest commands, we only need to detail them, it makes sense to  remove all of them if course

Regarding the branch name I've not come up with a working regexp to tell the branch name from the file path, as the branch name can contain slashes, more than happy to strip the branch from path and expose it as a variable

**Comment by yakky** [comment]
@jayvdb I have been able to look at this only after so much time, unfortunately

But I spent some time to add some git-url-parse compatibility

File: giturlparse/result.py (hunk 2 above)
Comment (yakky) [reply giturlparse/result.py]: `pathname` is now the same as `git-url-parse`, but I could not find a way to get the branch for the `blob` verb

`tree` verb instead can return the branch name

**Comment by yakky** [comment]
@jayvdb this has been replaced by #26 for rebase against current master

**Pull request closed without merging**